connector.go

```diff
@@ -166,6 +190,104 @@ func (c *connector) Start(ctx context.Context) {
 	logger.Debug("cancel channel triggered")
 }
 
+func (c *connector) shouldTakeSnapshot(ctx context.Context) bool {
+	if !c.cfg.Snapshot.Enabled {
+		return false
+	}
+
+	switch c.cfg.Snapshot.Mode {
+	case config.SnapshotModeNever:
+		return false
+	case config.SnapshotModeInitial:
+		state, err := c.snapshotter.LoadState(ctx, c.cfg.Slot.Name)
+		if err != nil {
+			logger.Debug("failed to load snapshot state, will take snapshot", "error", err)
+			return true
+		}
+		return state == nil || !state.Completed
+	default:
+		logger.Warn("invalid snapshot mode, skipping snapshot", "mode", c.cfg.Snapshot.Mode)
+		return false
+	}
+}
+
+func (c *connector) takeSnapshotWithRetry(ctx context.Context) error {
+	logger.Info("taking initial snapshot...")
+
+	// Create timeout context
+	timeoutCtx, cancel := context.WithTimeout(ctx, c.cfg.Snapshot.Timeout)
+	defer cancel()
+
+	var lastErr error
+
+	maxRetries := c.cfg.Snapshot.MaxRetries
+	for attempt := 1; attempt <= maxRetries; attempt++ {
+		logger.Info("snapshot attempt", "attempt", attempt, "maxRetries", maxRetries)
+
+		// Execute snapshot
+		err := c.snapshotter.TakeSnapshot(timeoutCtx, c.snapshotHandler, c.cfg.Slot.Name)
+		if err == nil {
+			logger.Info("snapshot completed successfully")
+
+			// Load existing state (from last checkpoint)
+			existingState, loadErr := c.snapshotter.LoadState(ctx, c.cfg.Slot.Name)
+			if loadErr != nil {
+				logger.Warn("failed to load final state", "error", loadErr)
+			}
+
+			// Mark snapshot as completed, preserving existing state
+			finalState := &snapshot.SnapshotState{
+				SlotName:        c.cfg.Slot.Name,
+				LastSnapshotLSN: c.system.LoadXLogPos(),
+				LastSnapshotAt:  time.Now().UTC(),
+				Completed:       true,
+			}
+
+			// Preserve checkpoint data if exists
+			if existingState != nil {
+				finalState.CurrentTable = existingState.CurrentTable
+				finalState.CurrentOffset = existingState.CurrentOffset
+				finalState.TotalRows = existingState.TotalRows
+			}
+
+			if saveErr := c.snapshotter.SaveState(ctx, finalState); saveErr != nil {
+				logger.Warn("failed to mark snapshot as completed", "error", saveErr)
```

Comment:
if we didn't  save or retry the state error, wouldn't this lead to the processing of duplicate initial data?